File: c++/202. Happy Number.cpp
class Solution {
public:
    bool isHappy(int n) {
        unordered_map<int, int> um;
        int t;
        while (n != 1 && um.find(n) == um.end()) {
            um[n]++;
            t = 0;
            while (n != 0) {
                t += (n%10)*(n%10);
                n /= 10;
            }
            
            n = t;
        }
        
        return(n==1);
    }
};

File: python/202. Happy Number.py
class Solution(object):
    def isHappy(self, n):
        """
        :type n: int
        :rtype: bool
        """
        s = set([n])
        while n!=1:
            nums = list(str(n))
            t = 0
            for num in nums:
                t += int(num)*int(num)
            if t in s: return False
            n = t
        return True

if __name__ == '__main__':
    su = Solution()
    su.isHappy(2)        